### PR TITLE
Add support for creating external resources associated with self-managed certificates

### PR DESCRIPTION
**Pull request by impl**
## Description

This change adds attributes to `auth0_custom_domain` and `auth0_custom_domain_verification` to make it possible to seamlessly manage all the additional infrastructure associated with self-managed certificates. I also added a complete example of setting up a custom domain using self-managed certs with GCP.

Dependencies:
* auth0/go-auth0#37

~~For testing purposes I updated the `go.mod` for this branch to point to our `go-auth0` fork. When the dependency lands, I'll update this PR to remove the temporary `replace` directive. (Please don't merge until then! :-)~~ Done!

## Checklist

**Note:** Checklist required to be completed before a PR is considered to be reviewable.

#### Auth0 Code of Conduct
- [x] I have read and agreed to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)

#### Auth0 General Contribution Guidelines
- [x] I have read the [Auth0 General Contribution Guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)

#### Changes include test coverage?
- [ ] Yes
- [x] Not needed

#### Does the description provide the correct amount of context?
- [x] Yes, the description provides enough context for the reviewer to understand what these changes accomplish

#### Have you updated the documentation?
- [x] Yes, I've updated the appropriate docs
- [ ] Not needed

#### Is this code ready for production?
- [x] Yes, all code changes are intentional and no debugging calls are left over

### CODE REVIEW

File: auth0/resource_auth0_custom_domain_verification.go
Comment (impl) [inline auth0/resource_auth0_custom_domain_verification.go:68]: Does this need to be wrapped in `d.Partial(true)`? I can't really figure out what the intent of that API is, but it would be _nice_ to not lose this key if the subsequent read fails for some transient reason.

Comment (sergiughf) [reply auth0/resource_auth0_custom_domain_verification.go:68]: Hey @impl 👋🏻  That's correct, there's no need for a `d.Partial(true)` over here. It's mostly a legacy function that was used together with `d.SetPartial("field")`, but honestly it was having a very weird behavior most of the times and considering terraform always refreshes before each operation it was really not needed.

File: auth0/resource_auth0_custom_domain_verification.go
Comment (sergiughf) [inline auth0/resource_auth0_custom_domain_verification.go:67]: Thanks for the context over here, super important to keep trace of this 👍🏻

File: auth0/resource_auth0_custom_domain_verification.go
Comment (sergiughf) [inline auth0/resource_auth0_custom_domain_verification.go:78]: This is a really good improvement, all creation should follow with a read. Thanks!

File: example/custom_domain_self_managed_certs/README.md
Comment (sergiughf) [inline example/custom_domain_self_managed_certs/README.md:1]: Incredibly thorough example, this is so much appreciated! Thanks @impl for always going the extra mile.

**Review by sergiughf — CHANGES_REQUESTED** [review]
Solid contribution @impl as always. There's just one thing I'd like us to add. 

Considering the custom domain verification always relies on a 3rd party (in your examples you introduced the google provider) we introduced a wiremock server to be able to run the tests for domain verification. You can spin it up from the root of the project using `docker compose` or running the `make dev-up` command. 

The response/request mappings are found here: https://github.com/auth0/terraform-provider-auth0/blob/main/dockerfiles/wiremock/__files/custom_domain_with_pending_verification.json. 

Would you be able to add an assertion for the 2 properties we introduced on the domain verification tests and update the request/response payloads as well to be able to parse them? 🙏🏻

File: auth0/resource_auth0_custom_domain_verification.go
Comment (impl) [reply auth0/resource_auth0_custom_domain_verification.go:78]: This was actually happening inside the retry loop before, but I moved it outside because any potential error was being masked.

**Comment by sergiughf** [comment]
@impl Let me know if you need any help with that, as we haven't really documented well that part in our contributing section. I'll update that next week for other folks as well to give better clarity. In the future we're aiming at giving the possibility for forks to run all the tests without requiring a real test tenant (and they'll run automatically on fork PRs), but having all requests / responses mocked. The actual Auth0 Management API should get hit only once we merge to master just to rule out any potential miss-alignments between the mocks. All of this will require some work from our end tho so I don't have an ETA when it will be available in full. For now we just use mocks for the domain verification.

**Comment by impl** [comment]
@sergiughf, that sounds like a great plan to me. I've been able to run relevant tests against a tenant I created for acceptance testing, but of course it would also be convenient to not have to do that in some cases. I went ahead and added a test for self-managed certificates and updated some of the Auth0-managed certificates test as well. To make the WireMock files a little less confusing I ended up doing some minor refactoring, which is entirely separate in [one commit](https://github.com/auth0/terraform-provider-auth0/pull/118/commits/86a763b9054b725092a6b2b8f515a815bd8186bc) in case there's anything specific there you'd like changed. Hopefully it will be enough of a bridge until you can get it set up for everything. Let me know!

File: Makefile
Comment (impl) [inline Makefile:107]: This fixes a test bug where running the acceptance tests multiple times in a row required a `dev-down` and `dev-up` cycle.

File: Makefile
Comment (impl) [inline Makefile:122]: Now that these files are exposed as volumes, you can simply edit the files locally and when you rerun the tests they will automatically be synced.

File: dockerfiles/wiremock/__files/custom_domain_verification/cd_auth0managed.json
Comment (impl) [inline dockerfiles/wiremock/__files/custom_domain_verification/cd_auth0managed.json:7]: I enabled response templating to make this work, but I could see it being confusing -- let me know if you want me to back this out and go back to separate files for each status.

File: Makefile
Comment (sergiughf) [reply Makefile:107]: First of all, I'd like to really commend the great work you've put into this PR and the follow up to make sure our testing is thorough. 🥳  You've really went above and beyond and we can't thank you enough. 

I generally like the improvements a lot however I encounter a few hiccups while reviewing and testing this locally. Let me explain:

- Everything works great if I always make sure to run all the tests through the `make test-acc` command
- However if I only run `TestAccCustomDomainVerificationWithSelfManagedCerts` or `TestAccCustomDomainVerificationWithAuth0ManagedCerts` at the second consecutive run I get `Error: invalid character 'R' looking for beginning of value` because now we're required to reset the scenarios before every run. So I have to make sure not to forget to do `docker-compose exec -- auth0 curl -4 -sS -o /dev/null -X POST http://localhost:8080/__admin/scenarios/reset`
- Before these changes I ran the previous test we had in a loop 100 times and it never failed without resetting
-  So I'm wondering what can we do in order for us to have the same behavior as before and not need to reset the scenarios after each test as it gets to be quite a cumbersome task to do?

File: dockerfiles/wiremock/__files/custom_domain_verification/cd_auth0managed.json
Comment (impl) [reply dockerfiles/wiremock/__files/custom_domain_verification/cd_auth0managed.json:7]: Yeah, that's what I was worried about too. I'll split it back out into a couple of files.

File: Makefile
Comment (impl) [reply Makefile:107]: So, interestingly enough, previously this worked only coincidentally. The attribute checks in the test were valid against both the "Started" and "Verified" responses. The first time you ran the test, it would correctly do the scenario state transition from the "Started" state, and then after that every test would just run against the "Verified" responses from the outset. You can validate that by checking the `status` attribute in the test.

I changed the transitions so that they have to follow the create-to-verify path, which is what's breaking now.

I was using `make test-acc TESTARGS="-run '.*TestAccCustomDomainVerification.*'"`, but I didn't test with `go test` directly. However that use case makes sense, and I think what we can actually do there is have each test reset the specific scenarios it cares about before running. Let me do a little work on that and get back to you!

Comment (impl) [reply Makefile]: The feature on the WireMock side to reset specific scenarios is new and their Docker image is broken or hasn't been updated yet (not sure what's going on there), so I opted to create a little client that the tests will simply call once to reset all mappings/scenarios the first time a mock provider is instantiated. That should give parity between calling `go test` in the Makefile and standalone. Let me know what you think!

Comment (sergiughf) [reply Makefile]: You're totally right @impl, we weren't checking for the status attribute and indeed before a subsequent call would make the test fail with that assertion present. Really appreciate you spotting that and fixing it! 🙇🏻 

I reviewed the PR and really like the approach 👏🏻 , as I wanted to save you some time I went ahead and refactored slightly the wiremock client package to remove the 3rd party retry sdk as IMO it's unnecessary. As I couldn't push on top of this PR as you had the "enable maintainers to modify contribution" checkbox thingie disabled I forked your fork and appended the changes over there https://github.com/puppetlabs/terraform-provider-auth0/pull/1. Let me know as well what you think:) and we can merge that into this PR or you can enable the checkbox on this PR so I can push on top, up to you. 

Thanks a lot for all the hard work put into this! It's a really great contribution.

**Comment by impl** [comment]
Of course, and thank you for the review/improvements and for being so responsive to these PRs! Our Terraform config looks better by the day ;).

**Pull request merged**
